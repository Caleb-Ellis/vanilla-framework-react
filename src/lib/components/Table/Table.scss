@import 'vanilla-framework/scss/base';
@import 'vanilla-framework/scss/patterns_table-sortable';
@import 'vanilla-framework/scss/utilities_content-align';
@include vf-b-typography;
@include vf-b-tables;
@include vf-p-table-sortable;
@include vf-u-align;
