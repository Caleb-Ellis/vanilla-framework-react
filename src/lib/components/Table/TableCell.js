import React from 'react';
import PropTypes from 'prop-types';
import getClassName from '../../utils/getClassName';

const TableCell = (props) => {
  const {
    align, ariaSort, children, className, columnHeader, role, rowHeader, ...otherProps
  } = props;

  const classNames = getClassName({
    [className]: className,
    [`u-align--${align}`]: align,
  }) || undefined;

  if (columnHeader) {
    return (
      <th
        scope="col"
        role="columnheader"
        aria-sort={ariaSort}
        className={classNames}
        {...otherProps}
      >
        {children}
      </th>);
  }

  if (rowHeader) {
    return (
      <th
        scope="row"
        role="rowheader"
        className={classNames}
        {...otherProps}
      >
        {children}
      </th>);
  }

  return (
    <td
      role={role}
      className={classNames}
      {...otherProps}
    >
      {children}
    </td>
  );
};

TableCell.defaultProps = {
  align: null,
  ariaSort: 'none',
  children: null,
  className: undefined,
  columnHeader: false,
  role: 'gridcell',
  rowHeader: false,
};

TableCell.propTypes = {
  align: PropTypes.string,
  ariaSort: PropTypes.string,
  children: PropTypes.node,
  className: PropTypes.string,
  columnHeader: PropTypes.bool,
  role: PropTypes.string,
  rowHeader: PropTypes.bool,
};

TableCell.displayName = 'TableCell';

export default TableCell;
