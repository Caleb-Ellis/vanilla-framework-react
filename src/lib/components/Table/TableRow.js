import React from 'react';
import PropTypes from 'prop-types';

const TableRow = (props) => {
  const { children, ...otherProps } = props;

  return (
    <tr role="row" {...otherProps}>
      {children}
    </tr>
  );
};

TableRow.propTypes = {
  children: PropTypes.node.isRequired,
};

TableRow.displayName = 'TableRow';

export default TableRow;
