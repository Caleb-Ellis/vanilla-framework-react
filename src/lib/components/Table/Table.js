import React from 'react';
import PropTypes from 'prop-types';
import naturalCompare from 'natural-compare-lite';
import getClassName from '../../utils/getClassName';

import TableCell from './TableCell';
import TableRow from './TableRow';
import Button from '../Button/Button';

class Table extends React.Component {
  constructor(props) {
    super(props);

    this.generateData = this.generateData.bind(this);
    this.generateHeader = this.generateHeader.bind(this);
    this.generateRows = this.generateRows.bind(this);
    this.generateCells = this.generateCells.bind(this);
    this.generateExpandedCell = this.generateExpandedCell.bind(this);
    this.getNewSortCondition = this.getNewSortCondition.bind(this);
    this.handleExpand = this.handleExpand.bind(this);
    this.sortData = this.sortData.bind(this);

    this.state = {
      columns: props.columns ? [...props.columns] : null,
      data: props.data ? [...props.data] : null,
      expandedRows: [],
      originalData: props.data ? [...props.data] : null,
      sortCondition: props.sortCondition,
    };
  }

  componentWillMount() {
    const { columns, data, sortCondition } = this.state;

    // If Table is static, convert data to object
    if (!data || !columns) {
      this.generateData();
    }

    // Pre-sort Table if sortCondition prop is non-zero
    if (Object.keys(sortCondition).length === 1) {
      const columnKey = Object.keys(sortCondition)[0];
      this.sortData(columnKey);
    }
  }

  componentWillReceiveProps(nextProps) {
    if (nextProps.data !== this.props.data) {
      this.setState({ data: [...nextProps.data], columns: [...nextProps.columns] });
    }
  }

  getNewSortCondition(columnKey) {
    const { sortCondition } = this.state;
    const newSortCondition = {};
    const options = ['ascending', 'descending', 'none'];

    // Move to next option in options array
    newSortCondition[columnKey] = (
      options[(options.indexOf(sortCondition[columnKey]) + 1) % options.length]
    );

    this.setState({ sortCondition: newSortCondition }, () => {
      this.sortData(columnKey);
    });
  }

  sortData(columnKey) {
    const { data, sortCondition } = this.state;

    if (sortCondition[columnKey] === 'none') {
      const unsortedData = this.state.originalData;
      this.setState({ data: [...unsortedData] });
    } else {
      const sortedData = data.sort((a, b) => {
        const x = a[columnKey];
        const y = b[columnKey];
        return (
          sortCondition[columnKey] === 'ascending' ? naturalCompare(x, y) : naturalCompare(y, x)
        );
      });

      this.setState({ data: sortedData });
    }
  }

  handleExpand(rowId) {
    const { accordion, expandable } = this.props;

    if (expandable) {
      const { expandedRows } = this.state;
      const isExpanded = expandedRows.includes(rowId);
      let newExpandedRows;

      if (isExpanded) {
        newExpandedRows = expandedRows.filter(id => id !== rowId);
      } else if (accordion) {
        newExpandedRows = [rowId];
      } else {
        newExpandedRows = expandedRows.concat(rowId);
      }

      this.setState({ expandedRows: newExpandedRows });
    }
  }

  generateData() {
    const { children } = this.props;
    const columns = [];
    const data = [];

    children.forEach((row, rowIndex) => {
      if (row.props.tableHeader) {
        row.props.children.forEach((cell, cellIndex) => {
          let value = '';
          if (typeof cell.props.children === 'string') value = cell.props.children;
          columns.push({
            key: `column${cellIndex}`,
            label: value,
            align: cell.props.align,
            header: cell.props.columnHeader,
          });
        });
      } else {
        const index = rowIndex - 1;
        data[index] = {};
        data[index].id = rowIndex;
        row.props.children.forEach((cell, cellIndex) => {
          let value = '';
          if (typeof cell.props.children === 'string') {
            value = cell.props.children;
          }
          if (cell.props.expandedCell) {
            data[index].expandedCell = cell.props.children;
          } else {
            data[index][`column${cellIndex}`] = value;
          }
        });
      }
    });

    this.setState({
      columns,
      data: [...data],
      originalData: data,
    });
  }

  generateHeader() {
    const { columns, sortCondition } = this.state;
    const { expandable, sortable } = this.props;

    let columnHeaders = columns.map(column => (
      <TableCell
        columnHeader
        key={column.key}
        ariaSort={sortCondition[column.key]}
        align={column.align}
        onClick={sortable ? () => this.getNewSortCondition(column.key) : null}
      >
        {column.label}
      </TableCell>),
    );

    if (expandable) {
      columnHeaders = columnHeaders.concat([
        <TableCell
          columnHeader
          key="expandCol"
          align="right"
          style={{ flexBasis: '4rem', flexGrow: '0' }}
          ariaSort={null}
        />,
        <TableCell className="u-hide" key="nullHeader" />,
      ]);
    }

    const header = (
      <thead>
        <TableRow key="header-row">
          {columnHeaders}
        </TableRow>
      </thead>
    );

    return header;
  }

  generateRows() {
    const { data } = this.state;
    const { expandable } = this.props;
    const rows = data.map((row) => {
      const cells = this.generateCells(row);

      return (
        <TableRow key={`row-${row.id}`}>
          {cells}
          {expandable && this.generateExpandedCell(row.id)}
        </TableRow>
      );
    });

    return rows;
  }

  generateCells(row) {
    const { columns, expandedRows } = this.state;
    const { expandable } = this.props;
    const isExpandable = 'expandedCell' in row;
    const isExpanded = expandedRows.includes(row.id);

    let cells = columns.map(column => (
      <TableCell
        rowHeader={column.header}
        key={`${column.key}-row-${row.id}`}
        align={column.align}
      >
        {row[column.key]}
      </TableCell>),
    );

    if (expandable) {
      cells = cells.concat(
        <TableCell
          key={`expandable-cell-${row.id}`}
          align="right"
          style={{ flexBasis: '4rem', flexGrow: '0' }}
        >
          {isExpandable ?
            <Button style={{ padding: '0' }} onClick={() => this.handleExpand(row.id)}>
              <i className={isExpanded ? 'p-icon--minus' : 'p-icon--plus'} />
            </Button> :
            ''
          }
        </TableCell>,
      );
    }

    return cells;
  }

  generateExpandedCell(rowId) {
    const { data, expandedRows } = this.state;
    const { expandedCell } = data.find(obj => obj.id === rowId);
    const isExpanded = expandedRows.includes(rowId);

    const classNames = getClassName({
      'p-table-expanding__panel': true,
      'u-hide': !expandedCell || !isExpanded,
    }) || undefined;

    return (
      <TableCell
        expandedCell
        style={{ borderTop: '0' }}
        className={classNames}
        aria-hidden={!expandedCell || !isExpanded}
      >
        {expandedCell}
      </TableCell>
    );
  }

  render() {
    const {
      className, expandable, hasHeader, sortable,
    } = this.props;

    const classNames = getClassName({
      [className]: className,
      'p-table-expanding': expandable,
      'p-table--sortable': sortable,
    }) || undefined;

    return (
      <table role="grid" className={classNames}>
        {hasHeader && this.generateHeader()}
        <tbody>
          {this.generateRows()}
        </tbody>
      </table>
    );
  }
}

Table.defaultProps = {
  accordion: false,
  children: null,
  className: undefined,
  columns: null,
  data: null,
  expandable: false,
  hasHeader: true,
  sortable: false,
  sortCondition: {},
};

Table.propTypes = {
  accordion: PropTypes.bool,
  children: PropTypes.node,
  className: PropTypes.string,
  columns: PropTypes.arrayOf(PropTypes.object),
  data: PropTypes.arrayOf(PropTypes.object),
  expandable: PropTypes.bool,
  hasHeader: PropTypes.bool,
  sortable: PropTypes.bool,
  sortCondition: PropTypes.object,
};

Table.displayName = 'Table';

export default Table;
