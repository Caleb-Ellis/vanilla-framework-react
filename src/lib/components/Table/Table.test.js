import React from 'react';
import ReactTestRenderer from 'react-test-renderer';
import Table from './Table';
import TableCell from './TableCell';
import TableRow from './TableRow';

const columns = [
  [
    { key: 'header', label: '', header: true },
    { key: 'column1', label: 'Header' },
    { key: 'column2', label: 'Header' },
    { key: 'column3', label: 'Header', align: 'right' },
  ],
  [
    { key: 'name', label: 'Name' },
    { key: 'users', label: 'Users', align: 'right' },
    { key: 'units', label: 'Units', align: 'right' },
    { key: 'revenue', label: 'Revenue', align: 'right' },
  ],
  [
    { key: 'username', label: 'Username' },
    { key: 'email', label: 'Email' },
    { key: 'height', label: 'Height (cm)', align: 'right' },
  ],
];

const data = [
  [{
    id: 0, header: 'Header row', column1: 'Table cell', column2: 'Table cell', column3: 111,
  }, {
    id: 1, header: 'Header row', column1: 'Table cell', column2: 'Table cell', column3: 222,
  }, {
    id: 2, header: 'Header row', column1: 'Table cell', column2: 'Table cell', column3: 333,
  }],
  [{
    id: 0, name: 'Grape', users: 8, units: 19, revenue: '$70',
  }, {
    id: 1, name: 'Melon', users: 12, units: 23, revenue: '$99',
  }, {
    id: 2, name: 'Apple', users: 9, units: 17, revenue: '$120',
  }],
  [{
    id: 0,
    username: 'aeinstein',
    email: 'albert.einstein@yahoo.com',
    height: 175,
    expandedCell: 'Hello',
  }, {
    id: 1,
    username: 'shawking',
    email: 'stephen.hawking@hotmail.com',
    height: 183,
    expandedCell: 'Hello',
  }, {
    id: 2,
    username: 'mcurie',
    email: 'marie.curie@gmail.com',
    height: 158,
    expandedCell: 'Hello',
  }],
];

describe('<Table>', () => {
  it('should render a static table correctly', () => {
    const table = ReactTestRenderer.create(
      <Table>
        <TableRow tableHeader>
          <TableCell columnHeader />
          <TableCell>Header</TableCell>
          <TableCell>Header</TableCell>
          <TableCell align="right">Header</TableCell>
        </TableRow>
        <TableRow>
          <TableCell>Header row</TableCell>
          <TableCell>Table cell</TableCell>
          <TableCell>Table cell</TableCell>
          <TableCell>111</TableCell>
        </TableRow>
        <TableRow>
          <TableCell>Header row</TableCell>
          <TableCell>Table cell</TableCell>
          <TableCell>Table cell</TableCell>
          <TableCell>222</TableCell>
        </TableRow>
        <TableRow>
          <TableCell>Header row</TableCell>
          <TableCell>Table cell</TableCell>
          <TableCell>Table cell</TableCell>
          <TableCell>333</TableCell>
        </TableRow>
      </Table>,
    );
    const json = table.toJSON();
    expect(json).toMatchSnapshot();
  });

  it('should render a dynamic table correctly', () => {
    const table = ReactTestRenderer.create(
      <Table
        columns={columns[0]}
        data={data[0]}
      />,
    );
    const json = table.toJSON();
    expect(json).toMatchSnapshot();
  });

  it('should render a headerless table correctly', () => {
    const table = ReactTestRenderer.create(
      <Table
        hasHeader={false}
        columns={columns[0]}
        data={data[0]}
      />,
    );
    const json = table.toJSON();
    expect(json).toMatchSnapshot();
  });

  it('should render a sortable table correctly', () => {
    const table = ReactTestRenderer.create(
      <Table
        sortable
        columns={columns[1]}
        data={data[1]}
      />,
    );
    const json = table.toJSON();
    expect(json).toMatchSnapshot();
  });

  it('should render a pre-sorted table correctly', () => {
    const table = ReactTestRenderer.create(
      <Table
        sortCondition={{ units: 'descending' }}
        columns={columns[1]}
        data={data[1]}
      />,
    );
    const json = table.toJSON();
    expect(json).toMatchSnapshot();
  });

  it('should render an expandable table correctly', () => {
    const table = ReactTestRenderer.create(
      <Table
        expandable
        columns={columns[2]}
        data={data[2]}
      />,
    );
    const json = table.toJSON();
    expect(json).toMatchSnapshot();
  });
});
