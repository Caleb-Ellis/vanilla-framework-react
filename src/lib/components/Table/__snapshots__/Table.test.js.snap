// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Table> should render a dynamic table correctly 1`] = `
<table
  className={undefined}
  role="grid"
>
  <thead>
    <tr
      role="row"
    >
      <th
        aria-sort="none"
        className={undefined}
        onClick={null}
        role="columnheader"
        scope="col"
      >
        
      </th>
      <th
        aria-sort="none"
        className={undefined}
        onClick={null}
        role="columnheader"
        scope="col"
      >
        Header
      </th>
      <th
        aria-sort="none"
        className={undefined}
        onClick={null}
        role="columnheader"
        scope="col"
      >
        Header
      </th>
      <th
        aria-sort="none"
        className="u-align--right"
        onClick={null}
        role="columnheader"
        scope="col"
      >
        Header
      </th>
    </tr>
  </thead>
  <tbody>
    <tr
      role="row"
    >
      <th
        className={undefined}
        role="rowheader"
        scope="row"
      >
        Header row
      </th>
      <td
        className={undefined}
        role="gridcell"
      >
        Table cell
      </td>
      <td
        className={undefined}
        role="gridcell"
      >
        Table cell
      </td>
      <td
        className="u-align--right"
        role="gridcell"
      >
        111
      </td>
    </tr>
    <tr
      role="row"
    >
      <th
        className={undefined}
        role="rowheader"
        scope="row"
      >
        Header row
      </th>
      <td
        className={undefined}
        role="gridcell"
      >
        Table cell
      </td>
      <td
        className={undefined}
        role="gridcell"
      >
        Table cell
      </td>
      <td
        className="u-align--right"
        role="gridcell"
      >
        222
      </td>
    </tr>
    <tr
      role="row"
    >
      <th
        className={undefined}
        role="rowheader"
        scope="row"
      >
        Header row
      </th>
      <td
        className={undefined}
        role="gridcell"
      >
        Table cell
      </td>
      <td
        className={undefined}
        role="gridcell"
      >
        Table cell
      </td>
      <td
        className="u-align--right"
        role="gridcell"
      >
        333
      </td>
    </tr>
  </tbody>
</table>
`;

exports[`<Table> should render a headerless table correctly 1`] = `
<table
  className={undefined}
  role="grid"
>
  <tbody>
    <tr
      role="row"
    >
      <th
        className={undefined}
        role="rowheader"
        scope="row"
      >
        Header row
      </th>
      <td
        className={undefined}
        role="gridcell"
      >
        Table cell
      </td>
      <td
        className={undefined}
        role="gridcell"
      >
        Table cell
      </td>
      <td
        className="u-align--right"
        role="gridcell"
      >
        111
      </td>
    </tr>
    <tr
      role="row"
    >
      <th
        className={undefined}
        role="rowheader"
        scope="row"
      >
        Header row
      </th>
      <td
        className={undefined}
        role="gridcell"
      >
        Table cell
      </td>
      <td
        className={undefined}
        role="gridcell"
      >
        Table cell
      </td>
      <td
        className="u-align--right"
        role="gridcell"
      >
        222
      </td>
    </tr>
    <tr
      role="row"
    >
      <th
        className={undefined}
        role="rowheader"
        scope="row"
      >
        Header row
      </th>
      <td
        className={undefined}
        role="gridcell"
      >
        Table cell
      </td>
      <td
        className={undefined}
        role="gridcell"
      >
        Table cell
      </td>
      <td
        className="u-align--right"
        role="gridcell"
      >
        333
      </td>
    </tr>
  </tbody>
</table>
`;

exports[`<Table> should render a pre-sorted table correctly 1`] = `
<table
  className={undefined}
  role="grid"
>
  <thead>
    <tr
      role="row"
    >
      <th
        aria-sort="none"
        className={undefined}
        onClick={null}
        role="columnheader"
        scope="col"
      >
        Name
      </th>
      <th
        aria-sort="none"
        className="u-align--right"
        onClick={null}
        role="columnheader"
        scope="col"
      >
        Users
      </th>
      <th
        aria-sort="descending"
        className="u-align--right"
        onClick={null}
        role="columnheader"
        scope="col"
      >
        Units
      </th>
      <th
        aria-sort="none"
        className="u-align--right"
        onClick={null}
        role="columnheader"
        scope="col"
      >
        Revenue
      </th>
    </tr>
  </thead>
  <tbody>
    <tr
      role="row"
    >
      <td
        className={undefined}
        role="gridcell"
      >
        Melon
      </td>
      <td
        className="u-align--right"
        role="gridcell"
      >
        12
      </td>
      <td
        className="u-align--right"
        role="gridcell"
      >
        23
      </td>
      <td
        className="u-align--right"
        role="gridcell"
      >
        $99
      </td>
    </tr>
    <tr
      role="row"
    >
      <td
        className={undefined}
        role="gridcell"
      >
        Grape
      </td>
      <td
        className="u-align--right"
        role="gridcell"
      >
        8
      </td>
      <td
        className="u-align--right"
        role="gridcell"
      >
        19
      </td>
      <td
        className="u-align--right"
        role="gridcell"
      >
        $70
      </td>
    </tr>
    <tr
      role="row"
    >
      <td
        className={undefined}
        role="gridcell"
      >
        Apple
      </td>
      <td
        className="u-align--right"
        role="gridcell"
      >
        9
      </td>
      <td
        className="u-align--right"
        role="gridcell"
      >
        17
      </td>
      <td
        className="u-align--right"
        role="gridcell"
      >
        $120
      </td>
    </tr>
  </tbody>
</table>
`;

exports[`<Table> should render a sortable table correctly 1`] = `
<table
  className="p-table--sortable"
  role="grid"
>
  <thead>
    <tr
      role="row"
    >
      <th
        aria-sort="none"
        className={undefined}
        onClick={[Function]}
        role="columnheader"
        scope="col"
      >
        Name
      </th>
      <th
        aria-sort="none"
        className="u-align--right"
        onClick={[Function]}
        role="columnheader"
        scope="col"
      >
        Users
      </th>
      <th
        aria-sort="none"
        className="u-align--right"
        onClick={[Function]}
        role="columnheader"
        scope="col"
      >
        Units
      </th>
      <th
        aria-sort="none"
        className="u-align--right"
        onClick={[Function]}
        role="columnheader"
        scope="col"
      >
        Revenue
      </th>
    </tr>
  </thead>
  <tbody>
    <tr
      role="row"
    >
      <td
        className={undefined}
        role="gridcell"
      >
        Grape
      </td>
      <td
        className="u-align--right"
        role="gridcell"
      >
        8
      </td>
      <td
        className="u-align--right"
        role="gridcell"
      >
        19
      </td>
      <td
        className="u-align--right"
        role="gridcell"
      >
        $70
      </td>
    </tr>
    <tr
      role="row"
    >
      <td
        className={undefined}
        role="gridcell"
      >
        Melon
      </td>
      <td
        className="u-align--right"
        role="gridcell"
      >
        12
      </td>
      <td
        className="u-align--right"
        role="gridcell"
      >
        23
      </td>
      <td
        className="u-align--right"
        role="gridcell"
      >
        $99
      </td>
    </tr>
    <tr
      role="row"
    >
      <td
        className={undefined}
        role="gridcell"
      >
        Apple
      </td>
      <td
        className="u-align--right"
        role="gridcell"
      >
        9
      </td>
      <td
        className="u-align--right"
        role="gridcell"
      >
        17
      </td>
      <td
        className="u-align--right"
        role="gridcell"
      >
        $120
      </td>
    </tr>
  </tbody>
</table>
`;

exports[`<Table> should render a static table correctly 1`] = `
<table
  className={undefined}
  role="grid"
>
  <thead>
    <tr
      role="row"
    >
      <th
        aria-sort="none"
        className={undefined}
        onClick={null}
        role="columnheader"
        scope="col"
      >
        
      </th>
      <th
        aria-sort="none"
        className={undefined}
        onClick={null}
        role="columnheader"
        scope="col"
      >
        Header
      </th>
      <th
        aria-sort="none"
        className={undefined}
        onClick={null}
        role="columnheader"
        scope="col"
      >
        Header
      </th>
      <th
        aria-sort="none"
        className="u-align--right"
        onClick={null}
        role="columnheader"
        scope="col"
      >
        Header
      </th>
    </tr>
  </thead>
  <tbody>
    <tr
      role="row"
    >
      <th
        className={undefined}
        role="rowheader"
        scope="row"
      >
        Header row
      </th>
      <td
        className={undefined}
        role="gridcell"
      >
        Table cell
      </td>
      <td
        className={undefined}
        role="gridcell"
      >
        Table cell
      </td>
      <td
        className="u-align--right"
        role="gridcell"
      >
        111
      </td>
    </tr>
    <tr
      role="row"
    >
      <th
        className={undefined}
        role="rowheader"
        scope="row"
      >
        Header row
      </th>
      <td
        className={undefined}
        role="gridcell"
      >
        Table cell
      </td>
      <td
        className={undefined}
        role="gridcell"
      >
        Table cell
      </td>
      <td
        className="u-align--right"
        role="gridcell"
      >
        222
      </td>
    </tr>
    <tr
      role="row"
    >
      <th
        className={undefined}
        role="rowheader"
        scope="row"
      >
        Header row
      </th>
      <td
        className={undefined}
        role="gridcell"
      >
        Table cell
      </td>
      <td
        className={undefined}
        role="gridcell"
      >
        Table cell
      </td>
      <td
        className="u-align--right"
        role="gridcell"
      >
        333
      </td>
    </tr>
  </tbody>
</table>
`;

exports[`<Table> should render an expandable table correctly 1`] = `
<table
  className="p-table-expanding"
  role="grid"
>
  <thead>
    <tr
      role="row"
    >
      <th
        aria-sort="none"
        className={undefined}
        onClick={null}
        role="columnheader"
        scope="col"
      >
        Username
      </th>
      <th
        aria-sort="none"
        className={undefined}
        onClick={null}
        role="columnheader"
        scope="col"
      >
        Email
      </th>
      <th
        aria-sort="none"
        className="u-align--right"
        onClick={null}
        role="columnheader"
        scope="col"
      >
        Height (cm)
      </th>
      <th
        aria-sort={null}
        className="u-align--right"
        role="columnheader"
        scope="col"
        style={
          Object {
            "flexGrow": "0.5",
          }
        }
      >
        Expand
      </th>
      <td
        className="u-hide"
        role="gridcell"
      />
    </tr>
  </thead>
  <tbody>
    <tr
      role="row"
    >
      <td
        className={undefined}
        role="gridcell"
      >
        aeinstein
      </td>
      <td
        className={undefined}
        role="gridcell"
      >
        albert.einstein@yahoo.com
      </td>
      <td
        className="u-align--right"
        role="gridcell"
      >
        175
      </td>
      <td
        className="u-align--right"
        role="gridcell"
        style={
          Object {
            "flexGrow": "0.5",
          }
        }
      >
        <button
          className="p-button--base"
          disabled={false}
          href={undefined}
          onClick={[Function]}
          style={
            Object {
              "padding": "0.5rem",
            }
          }
        >
          <i
            className="p-icon--plus"
          />
        </button>
      </td>
      <td
        aria-hidden={true}
        className="p-table-expanding__panel u-hide"
        expandedCell={true}
        role="gridcell"
      >
        Hello
      </td>
    </tr>
    <tr
      role="row"
    >
      <td
        className={undefined}
        role="gridcell"
      >
        shawking
      </td>
      <td
        className={undefined}
        role="gridcell"
      >
        stephen.hawking@hotmail.com
      </td>
      <td
        className="u-align--right"
        role="gridcell"
      >
        183
      </td>
      <td
        className="u-align--right"
        role="gridcell"
        style={
          Object {
            "flexGrow": "0.5",
          }
        }
      >
        <button
          className="p-button--base"
          disabled={false}
          href={undefined}
          onClick={[Function]}
          style={
            Object {
              "padding": "0.5rem",
            }
          }
        >
          <i
            className="p-icon--plus"
          />
        </button>
      </td>
      <td
        aria-hidden={true}
        className="p-table-expanding__panel u-hide"
        expandedCell={true}
        role="gridcell"
      >
        Hello
      </td>
    </tr>
    <tr
      role="row"
    >
      <td
        className={undefined}
        role="gridcell"
      >
        mcurie
      </td>
      <td
        className={undefined}
        role="gridcell"
      >
        marie.curie@gmail.com
      </td>
      <td
        className="u-align--right"
        role="gridcell"
      >
        158
      </td>
      <td
        className="u-align--right"
        role="gridcell"
        style={
          Object {
            "flexGrow": "0.5",
          }
        }
      >
        <button
          className="p-button--base"
          disabled={false}
          href={undefined}
          onClick={[Function]}
          style={
            Object {
              "padding": "0.5rem",
            }
          }
        >
          <i
            className="p-icon--plus"
          />
        </button>
      </td>
      <td
        aria-hidden={true}
        className="p-table-expanding__panel u-hide"
        expandedCell={true}
        role="gridcell"
      >
        Hello
      </td>
    </tr>
  </tbody>
</table>
`;
